nestkernel/event_delivery_manager.h

```diff
@@ -52,6 +52,98 @@ typedef MPIManager::OffGridSpike OffGridSpike;
 class TargetData;
 class SendBufferPosition;
 
+/*
+ * A thread-safe array to coordinate progress across threads during
+ * gather operations.
+ */
+class CompletedChecker
+{
+private:
+  bool* a_;
+  size_t size_;
+
+public:
+  CompletedChecker()
+    : a_( NULL )
+    , size_( 0 )
+  {
+  }
+
+  ~CompletedChecker()
+  {
+    clear();
+  }
+
+  void clear()
+  {
+    if ( a_ != NULL )
+    {
+      delete a_;
+      a_ = NULL;
+
+      assert( size_ > 0 );
+      size_ = 0;
+    }
+  }
+
+  void resize( const size_t new_size, const bool v )
+  {
+    VPManager::assert_single_threaded();
+    clear();
+
+    a_ = new bool[ new_size ];
+    for ( size_t i = 0; i < new_size; ++i )
+    {
+      a_[ i ] = v;
+    }
+    size_ = new_size;
+  }
+
+  // sets element (thread safe)
+  void set( const thread tid, const bool v )
+  {
+#pragma omp atomic write
+    a_[ tid ] = v;
+  }
+
+  // updates element (thread safe)
+  void logical_and( const thread tid, const bool v )
+  {
+#pragma omp atomic update
+    a_[ tid ] = a_[ tid ] & v;
+  }
+
+  // returns whether all elements are true; waits for all threads
+  bool all_true() const
+  {
+    bool tmp = true;
+#pragma omp barrier
+    for ( size_t i = 0; i < size_; ++i )
+    {
+      tmp &= a_[ i ];
```

Comment:
`&=`  applies a *bitwise*  and, while this is a logical and, at least logically. To avoid subtle problems that could occur if someone decided to change the data type of `a_` to something else than bool, I would suggest `tmp = tmp and a_[i];`.

Another solution, possibly slightly faster, would be
```c++
for ( size_t i = 0 ; i < size_ ; ++i )
{
  if ( not a_[i] )
  {
    return false;
  }
}
return true;
```
without any need for a local variable.